In [1]:
# import pyttsx3
# text = pyttsx3.init()
# text.setProperty('rate', 150)
# text.setProperty('volume', 1) # 0 to 1
# s = '''DOMAIN: Digital content management
# • CONTEXT: Classification is probably the most popular task that you would deal with in real life. Text in the form of blogs, posts, articles, etc.
# are written every second. It is a challenge to predict the information about the writer without knowing about him/her. We are going to create a
# classifier that predicts multiple features of the author of a given text. We have designed it as a Multi label classification problem.
# • DATA DESCRIPTION: Over 600,000 posts from more than 19 thousand bloggers The Blog Authorship Corpus consists of the collected posts of
# 19,320 bloggers gathered from blogger.com in August 2004. The corpus incorporates a total of 681,288 posts and over 140 million words - or
# approximately 35 posts and 7250 words per person. Each blog is presented as a separate file, the name of which indicates a blogger id# and
# the blogger’s self-provided gender, age, industry, and astrological sign. (All are labelled for gender and age but for many, industry and/or sign is
# marked as unknown.) All bloggers included in the corpus fall into one of three age groups:
# • 8240 "10s" blogs (ages 13-17),
# • 8086 "20s" blogs(ages 23-27) and
# • 2994 "30s" blogs (ages 33-47)
# • For each age group, there is an equal number of male and female bloggers. Each blog in the corpus includes at least 200 occurrences of
# common English words. All formatting has been stripped with two exceptions. Individual posts within a single blogger are separated by the
# date of the following post and links within a post are denoted by the label url link.
# • PROJECT OBJECTIVE: To build a NLP classifier which can use input text parameters to determine the label/s of the blog. Specific to this case
# study, you can consider the text of the blog: ‘text’ feature as independent variable and ‘topic’ as dependent variable.
# Steps and tasks: [ Total Score: 40 Marks]
# 1. Read and Analyse Dataset. [5 Marks]
# A. Clearly write outcome of data analysis(Minimum 2 points) [2 Marks]
# B. Clean the Structured Data [3 Marks]
# i. Missing value analysis and imputation. [1 Marks]
# ii. Eliminate Non-English textual data. [2 Marks]
# Hint: Refer ‘langdetect’ library to detect language of the input text)
# 2. Preprocess unstructured data to make it consumable for model training. [5 Marks]
# A. Eliminate All special Characters and Numbers [2 Marks]
# B. Lowercase all textual data [1 Marks]
# C. Remove all Stopwords [1 Marks]
# D. Remove all extra white spaces [1 Marks]
# 3. Build a base Classification model [8 Marks]
# A. Create dependent and independent variables [2 Marks]
# Hint: Treat ‘topic’ as a Target variable.
# B. Split data into train and test. [1 Marks]
# C. Vectorize data using any one vectorizer. [2 Marks]
# D. Build a base model for Supervised Learning - Classification. [2 Marks]
# E. Clearly print Performance Metrics. [1 Marks]
# Hint: Accuracy, Precision, Recall, ROC-AUC
# 4. Improve Performance of model. [14 Marks]
# A. Experiment with other vectorisers. [4 Marks]
# B. Build classifier Models using other algorithms than base model. [4 Marks]
# C. Tune Parameters/Hyperparameters of the model/s. [4 Marks]
# D. Clearly print Performance Metrics. [2 Marks]
# Hint: Accuracy, Precision, Recall, ROC-AUC

# 5. Share insights on relative performance comparison [8 Marks]
# A. Which vectorizer performed better? Probable reason?.
# [2 Marks]
# B. Which model outperformed? Probable reason? [2 Marks]
# C. Which parameter/hyperparameter significantly helped
# to improve performance?Probable reason?. [2 Marks]
# D. According to you, which performance metric should be
# given most importance, why?. [2 Marks]'''
# text.say(s)
# text.runAndWait()

> # PART A - 40 Marks

### • DOMAIN: 
        Digital content management
### • CONTEXT:
        Classification is probably the most popular task that you would deal with in real life. Text in the form of blogs, posts, articles, etc. are written every second. It is a challenge to predict the information about the writer without knowing about him/her. We are going to create a classifier that predicts multiple features of the author of a given text. We have designed it as a Multi label classification problem. 
### • DATA DESCRIPTION: 
        Over 600,000 posts from more than 19 thousand bloggers The Blog Authorship Corpus consists of the collected posts of 19,320 bloggers gathered from blogger.com in August 2004. The corpus incorporates a total of 681,288 posts and over 140 million words - or approximately 35 posts and 7250 words per person. Each blog is presented as a separate file, the name of which indicates a blogger id# and the blogger’s self-provided gender, age, industry, and astrological sign. (All are labelled for gender and age but for many, industry and/or sign is
        marked as unknown.) All bloggers included in the corpus fall into one of three age groups:
            • 8240 "10s" blogs (ages 13-17),
            • 8086 "20s" blogs(ages 23-27) and
            • 2994 "30s" blogs (ages 33-47)
            • For each age group, there is an equal number of male and female bloggers. Each blog in the corpus includes at least 200 occurrences of common English words. All formatting has been stripped with two exceptions. Individual posts within a single blogger are separated by the date of the following post and links within a post are denoted by the label url link.
### • PROJECT OBJECTIVE: 
        To build a NLP classifier which can use input text parameters to determine the label/s of the blog. Specific to this case study, you can consider the text of the blog: ‘text’ feature as independent variable and ‘topic’ as dependent variable.

## Steps and tasks: [ Total Score: 40 Marks]

### 1. Read and Analyse Dataset. [5 Marks]

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

# import zipfile
# with zipfile.ZipFile('blogs.zip', 'r') as zip_ref:
#     zip_ref.extractall('E:\\RNN\\blogs')

In [2]:
import nltk
from nltk.tokenize.toktok import ToktokTokenizer
import re
from bs4 import BeautifulSoup
from contractions import contractions_dict
import unicodedata
from langdetect import detect
import spacy
nlp = spacy.load('en_core_web_sm')
tokenizer = ToktokTokenizer()
stopword_list = nltk.corpus.stopwords.words('english')


In [3]:
blogs  = pd.read_csv('E:\\RNN\\blogs.zip', compression='zip', header=0, sep=',', quotechar='"') # reading the content directly from zip file


# blogs = pd.read_csv('E:\\RNN\\blogs\\Output_1.csv')
blogs.sample(7)

id  gender  age                 topic       sign            date  \
556470  3071102  female   33            Technology     Cancer    27,July,2004   
578653  1578617    male   27  Communications-Media     Cancer    26,July,2004   
212727  3157701    male   15            Non-Profit        Leo    28,June,2004   
166920  3361994  female   24            Non-Profit  Capricorn     16,May,2004   
424593   579034  female   17               Student   Aquarius    29,July,2003   
73268    317581    male   26            Technology    Scorpio    20,June,2004   
6358     883178    male   36               Fashion      Aries  16,August,2002   

                                                     text  
556470             BUT, I KNOW I'M SPECIAL! Seriously,...  
578653         John C Wright, SF writer and semi-pro w...  
212727               Cat had kittens. big whoop. Perso...  
166920                   this should be fun, anybody h...  
424593         Hey all... every one has these post thi...  
73268          I'm going to pick up Cliff from the air...  
6358               Forced into exile by the crap calle...

        A. Clearly write outcome of data analysis(Minimum 2 points) [2 Marks]


In [4]:
blogs.shape

(681284, 7)

In [5]:
blogs.columns

Index(['id', 'gender', 'age', 'topic', 'sign', 'date', 'text'], dtype='object')

In [6]:
blogs.describe(include='all')

id  gender            age   topic    sign            date  \
count   6.812840e+05  681284  681284.000000  681284  681284          681284   
unique           NaN       2            NaN      40      12            2616   
top              NaN    male            NaN  indUnk  Cancer  02,August,2004   
freq             NaN  345193            NaN  251015   65048           16544   
mean    2.397802e+06     NaN      23.932326     NaN     NaN             NaN   
std     1.247723e+06     NaN       7.786009     NaN     NaN             NaN   
min     5.114000e+03     NaN      13.000000     NaN     NaN             NaN   
25%     1.239610e+06     NaN      17.000000     NaN     NaN             NaN   
50%     2.607577e+06     NaN      24.000000     NaN     NaN             NaN   
75%     3.525660e+06     NaN      26.000000     NaN     NaN             NaN   
max     4.337650e+06     NaN      48.000000     NaN     NaN             NaN   

                       text  
count                681284  
unique               611652  
top              urlLink     
freq                    445  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [7]:
blogs.dtypes

id         int64
gender    object
age        int64
topic     object
sign      object
date      object
text      object
dtype: object

### Outcome of Data Analysis

* There are 2 genders, 40 topics and 12 signs.
* When we look at unique and count for text, we can say there are 69632 (681284 - 611652) duplicate texts. 
* Among the duplicate blog text, urlLink is repeating mosltly.
* We have more male bloggers, most blogs belong to indUnk and most bloggers have cancer as their sign.
* There are no null values.


        B. Clean the Structured Data [3 Marks]
                i. Missing value analysis and imputation. [1 Marks]
                ii. Eliminate Non-English textual data. [2 Marks]
                Hint: Refer ‘langdetect’ library to detect language of the input text)

In [8]:
# Missing value 
blogs.isnull().sum()  # from the count in describe also we can say that there are no missing values.

id        0
gender    0
age       0
topic     0
sign      0
date      0
text      0
dtype: int64

In [9]:
# Eliminate Non-English textual data using langdetect. 
print("Shape of Blogs before eliminating Non-English Text", blogs.shape)

loc = 0
eng = 0
other =0
for blog in blogs['text']:
    
    try:
        if (detect(blog) == 'en'):
            eng = eng + 1
        else:
            # print(loc)
            blogs.drop(loc, inplace = True)
            other = other +1
            
            
    except:
        pass
    loc = loc +1



Shape of Blogs before eliminating Non-English Text (681284, 7)


In [28]:
print("Shape of Blogs after eleminating Non-English Text:", blogs.shape)
print("Number of English Blogs:", eng)
print("Number of other language Blogs which were removed:", other)

Shape of Blogs after eleminating Non-English Text: (655041, 3)
Number of English Blogs: 651489
Number of other language Blogs which were removed: 26243


In [11]:
blogs.sample(6)

id  gender  age           topic         sign              date  \
406605  3630901    male   34      Technology          Leo      20,July,2004   
149310  1405766    male   24  HumanResources      Scorpio    09,Aprill,2004   
364881  1912582  female   46          indUnk        Virgo      23,July,2004   
275393  3509187    male   16         Student  Sagittarius      06,June,2004   
483578   769349    male   34            Arts      Scorpio       30,May,2004   
195444  1951423  female   24            Arts      Scorpio  17,February,2004   

                                                     text  
406605         S and I were sitting on a park bench ta...  
149310         It's official - I propose a writing con...  
364881      urlLink    QuitNet certificate for 7 days ...  
275393     O wait one more thing...  Touching Evil: Aw...  
483578         All right, it's time to get down to som...  
195444         I suppose this is about the time where ...

In [12]:
blogs.shape

(655041, 7)

### 2. Preprocess unstructured data to make it consumable for model training. [5 Marks]
        A. Eliminate All special Characters and Numbers [2 Marks]

In [13]:
# def strip_html_tags(text):
#     soup = BeautifulSoup(text, "html.parser")
#     stripped_text = soup.get_text()
#     return stripped_text

# strip_html_tags('<html><h2>Some important text</h2></html>')

In [14]:
# Consider a sentence like (I'll be a Father). If we remove the special characters before contractions, then it becomes 'Ill be a Father'. Which actually changes the entire meaning.
# So lets apply contractions first and then remove special characters.

def expand_contractions(text,contraction_mapping=contractions_dict):
    contractions_pattern=re.compile('({})'.format('|'.join
                                      (contraction_mapping.keys())),flags=re.IGNORECASE|re.DOTALL)
    def expand_match(contraction):
        match=contraction.group(0)
        first_char=match[0]
        expanded_contraction=contraction_mapping.get(match)\
            if contraction_mapping.get(match)\
            else contraction_mapping.get(match.lower())
        expanded_contraction=first_char+expanded_contraction[1:]
        return expanded_contraction   
    
    
    try:                   
        expanded_text=contractions_pattern.sub(expand_match,str(text))
        expanded_text=re.sub("'"," ",expanded_text)
    except:
        return text
    return expanded_text

def remove_spl_char_n_numbers(text, remove_digits=True):
    # text = expand_contractions(text)
    pattern = r'[^a-zA-z0-9\s]' if not remove_digits else r'[^a-zA-z\s]'
    text = re.sub(pattern, '', text)
    return text
    

In [15]:
blogs.tail()

id gender  age    topic    sign          date  \
681279  1713845   male   23  Student  Taurus  01,July,2004   
681280  1713845   male   23  Student  Taurus  01,July,2004   
681281  1713845   male   23  Student  Taurus  01,July,2004   
681282  1713845   male   23  Student  Taurus  01,July,2004   
681283  1713845   male   23  Student  Taurus  01,July,2004   

                                                     text  
681279         Dear Susan,  I could write some really ...  
681280         Dear Susan,  'I have the second yeast i...  
681281         Dear Susan,  Your 'boyfriend' is fuckin...  
681282         Dear Susan:    Just to clarify, I am as...  
681283         Hey everybody...and Susan,  You might a...

In [16]:
# i = 0
# for blog in blogs['text']:
#     blog = remove_spl_char_n_numbers(blog)
    

blogs['text'] = blogs['text'].apply(expand_contractions)
blogs['text'] = blogs['text'].apply(remove_spl_char_n_numbers)
blogs.sample(18)

id  gender  age          topic         sign               date  \
9141    2581876  female   24         indUnk  Sagittarius       13,June,2004   
464081  2938013    male   23    Engineering    Capricorn       28,July,2004   
138194  1935335  female   25        Banking        Aries       27,June,2004   
355533    49663    male   33         indUnk       Taurus  27,September,2000   
358325  1280497    male   26     Technology       Taurus   05,November,2003   
88268    320317    male   36     Technology       Pisces      01,April,2002   
288926  3401120    male   15        Student       Cancer       06,June,2004   
635967  3306589    male   27        Science       Taurus       11,June,2004   
654390  4087530  female   16        Student       Taurus     02,August,2004   
222880  1477098  female   17      Marketing  Sagittarius      22,April,2004   
144807  2627816    male   13      Education        Libra        11,May,2004   
74075   2040365    male   26    Engineering    Capricorn    09,October,2003   
536230  1391636  female   26       Religion       Cancer  05,September,2002   
295985  3053026    male   17         indUnk     Aquarius       02,June,2004   
606762  1131517  female   17         indUnk       Gemini      22,April,2003   
476690  3647663  female   45  Manufacturing  Sagittarius       15,June,2004   
73465   2936892  female   14        Student       Cancer        22,May,2004   
403268  3952637  female   33     Non-Profit     Aquarius       20,July,2004   

                                                     text  
9141           filomena is the girl i want to be with ...  
464081               It is officialnbsp Jazz is my all...  
138194               Each week millions of Americans t...  
355533             Absolutely Fabulous fans click  url...  
358325                   I have to post a eulogy of so...  
88268               urlLink IBM Ease of Use What is us...  
288926             one of my older cousins Frank or Fr...  
635967         Great game  urlLink check out the dehem...  
654390         Momentarily listening to Lali Puna Faki...  
222880             I went to Penn Station to get an ap...  
144807                   I am UPDATING I am going to t...  
74075              wow  chk this out  do you know ther...  
536230         I went for a walk today  It rained  pou...  
295985               Thank you Mary for providing some...  
606762                   Today was a real drag I dont ...  
476690         Today I worked at Weber Aircraft from  ...  
73465              Hi back in here but in a terrible m...  
403268          urlLink    Who am I  Once upon a time ...

In [17]:
# blogs.drop(columns=["gender", 'age','sign','date','text_new'],inplace=True)
# blogs.drop(columns=['text_new2'],inplace=True)
# blogs.sample(20)

In [29]:
expand_contractions("i'll probably be moving to canada  ")

"i'll probably be moving to canada  "

In [19]:
# import unicodedata
# def remove_accented_chars(text):
#     text = unicodedata.normalize('NFKD', text).encode('ascii', 'ignore').decode('utf-8', 'ignore')
#     return text


In [20]:
# def simple_stemmer(text):
#     ps = nltk.porter.PorterStemmer()
#     text = ' '.join([ps.stem(word) for word in text.split()])
#     return text

# simple_stemmer("My system keeps crashing his crashed yesterday, ours crashes daily")



# def lemmatize_text(text):
#     text = nlp(text)
#     text = ' '.join([word.lemma_ if word.lemma_ != '-PRON-' else word.text for word in text])
#     return text

# lemmatize_text("My system keeps crashing! his crashed yesterday, ours crashes daily")


# def remove_stopwords(text, is_lower_case=False):
#     tokens = tokenizer.tokenize(text)
#     tokens = [token.strip() for token in tokens]
#     if is_lower_case:
#         filtered_tokens = [token for token in tokens if token not in stopword_list]
#     else:
#         filtered_tokens = [token for token in tokens if token.lower() not in stopword_list]
#     filtered_text = ' '.join(filtered_tokens)    
#     return filtered_text

# remove_stopwords("The, and, if are stopwords, computer is not")



        B. Lowercase all textual data [1 Marks]

In [21]:
blogs['text'] = blogs['text'].apply(lambda x:x.lower())
blogs.sample(4)

id  gender  age       topic         sign             date  \
451930  2983546  female   33      indUnk        Aries   03,August,2004   
540086  2020321  female   25      indUnk  Sagittarius      07,May,2004   
282024   958176    male   17  Non-Profit       Gemini  08,October,2003   
616958  3365547    male   17      indUnk      Scorpio      17,May,2004   

                                                     text  
451930             fine i admit it i am envious i envy...  
540086               chris  i saw van helsing after dr...  
282024                   that is very true  using my m...  
616958                deviant   urllink ssilence   my ...

        C. Remove all Stopwords [1 Marks]

In [22]:

def remove_stopwords(text, is_lower_case=False):
    tokens = tokenizer.tokenize(text)
    tokens = [token.strip() for token in tokens]
    if is_lower_case:
        filtered_tokens = [token for token in tokens if token not in stopword_list]
    else:
        filtered_tokens = [token for token in tokens if token.lower() not in stopword_list]
    filtered_text = ' '.join(filtered_tokens)    
    return filtered_text


blogs['text'] = blogs['text'].apply(remove_stopwords)
blogs.sample(4)

id  gender  age      topic      sign            date  \
144869  3820472  female   17    Student     Aries    17,July,2004   
26178   1939766  female   23       Arts  Aquarius    08,July,2004   
362100  3585348    male   25  Chemicals  Aquarius  09,August,2004   
185008  3971148  female   27       Arts       Leo    23,July,2004   

                                                     text  
144869  okie time special abit st thing coz early sund...  
26178   tomorrow big day quitting maybe quitting song ...  
362100   would got ehemail banksy sent one followed think  
185008  getting phone calls someone whose name recall ...

        
        D. Remove all extra white spaces [1 Marks]

In [23]:
def remove_whitespaces(text):
    text = re.sub(' +', ' ', text)
    return text

blogs['text'] = blogs['text'].apply(remove_whitespaces)
blogs.sample(10)

id  gender  age       topic      sign              date  \
115637  3403444  female   14      indUnk    Taurus      11,July,2004   
438765  2159339  female   26  Technology     Aries      03,July,2004   
642856  2608756  female   38   Education    Gemini     22,March,2004   
520742   898887    male   24      indUnk    Gemini       31,May,2003   
618672  1684840    male   15     Student   Scorpio    06,August,2003   
11712   3668449  female   24     Banking  Aquarius      29,June,2004   
208382  2681786  female   36     Fashion    Pisces    07,August,2004   
247729  4075390  female   24  Non-Profit    Pisces   27,October,2003   
368785  3579464  female   16     Student       Leo      09,July,2004   
85645   2259044  female   16     Student       Leo  22,December,2003   

                                                     text  
115637  fabityfabfab davida listen mesing fabityfabfab...  
438765  days nbsp days loafing around im taking rest w...  
642856  ive working another column bilingual newspaper...  
520742  shit didnt hit fan everything else know usuall...  
618672  woohoo got back city sweating like pig got loa...  
11712   urllink dnbsp urllink summarization english la...  
208382  urllink summer smith courtesy urllink bunny eh...  
247729  whilst driving work today flipping morning rad...  
368785  official worlds biggest flirt worked hour shif...  
85645   jello ive come strange conclusion jello u may ...

In [24]:
# blogs['text_new3'].str.replace("  ","")

### 3. Build a base Classification model [8 Marks]
        A. Create dependent and independent variables [2 Marks]
        Hint: Treat ‘topic’ as a Target variable.


In [25]:
# since X = blogs['text'] and y =['topic']]

x = blogs['text']

y = blogs['topic']

        B. Split data into train and test. [1 Marks]

In [26]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test =train_test_split(x,y, random_state=369)

        C. Vectorize data using any one vectorizer. [2 Marks]

In [30]:

from sklearn.feature_extraction.text import CountVectorizer,TfidfVectorizer

cv = CountVectorizer(min_df=10) # lowercase=True,stop_words='english' are handled already
print('x shape:',x_train.shape)
cv.fit_transform(x_train)

#Check the vocablury size
len(cv.vocabulary_)


x shape: (655041,)


102704

In [31]:
#Convert Training blog text into Count Vectors
x_train_ct = cv.transform(x_train)

#Size of Document Term Matrix
x_train_ct.shape

(491280, 102704)

In [32]:
# Apply transform on test data
x_test_ct = cv.transform(x_test)
x_test_ct.shape


(163761, 102704)

        D. Build a base model for Supervised Learning - Classification. [2 Marks]


In [33]:
from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB()
nb.fit(x_train_ct, y_train)
y_pred_class = nb.predict(x_test_ct)
print("Number of Features : ", x_train_ct.shape[1])
print("Training Accuracy : ",nb.score(x_train_ct,y_train ))
print("Testing Accuracy : ", nb.score(x_test_ct,y_test))

Number of Features :  102704
Training Accuracy :  0.4187103077674646
Testing Accuracy :  0.35957889851674085


        E. Clearly print Performance Metrics. [1 Marks]
        Hint: Accuracy, Precision, Recall, ROC-AUC

In [35]:
# Printing the confusion matrix
# print(confusion_matrix(y_test, y_pred_class))
# Printing the precision and recall, among other metrics
from sklearn.metrics import classification_report, confusion_matrix
print(classification_report(y_test, y_pred_class))

                         precision    recall  f1-score   support

             Accounting       0.41      0.21      0.28       897
            Advertising       0.27      0.04      0.07      1115
            Agriculture       0.22      0.01      0.01       287
           Architecture       0.16      0.11      0.13       387
                   Arts       0.21      0.22      0.22      7726
             Automotive       0.00      0.00      0.00       273
                Banking       0.19      0.04      0.07       973
                Biotech       0.30      0.05      0.09       552
       BusinessServices       0.50      0.13      0.21      1073
              Chemicals       0.14      0.04      0.07       914
   Communications-Media       0.16      0.15      0.15      4915
           Construction       0.06      0.01      0.01       249
             Consulting       0.17      0.04      0.07      1423
              Education       0.25      0.15      0.19      7206
            Engineering 

### 4. Improve Performance of model. [14 Marks]
        A. Experiment with other vectorisers. [4 Marks]



In [36]:
# Lets use Tfidf Vectorizer now
from sklearn.feature_extraction.text import TfidfVectorizer
tv = TfidfVectorizer(min_df=10)
tv.fit_transform(x_train)
#Check the vocablury size
len(tv.vocabulary_)

87083

In [37]:
#Convert Training blog text into Count Vectors
x_train_tv = tv.transform(x_train)

#Size of Document Term Matrix
x_train_tv.shape

(491280, 87083)

In [38]:
# Apply transform on test data
x_test_tv = tv.transform(x_test)
x_test_tv.shape


(163761, 87083)

        B. Build classifier Models using other algorithms than base model. [4 Marks]


In [39]:
from sklearn.ensemble import RandomForestClassifier
rfclf = RandomForestClassifier(max_depth=100)
rfclf.fit(x_train_tv, y_train)
y_pred_class = nb.predict(x_test_ct)
print("Number of Features : ", x_train_tv.shape[1])
print("Training Accuracy : ",rfclf.score(x_train_tv,y_train ))
print("Testing Accuracy : ", rfclf.score(x_test_tv,y_test))

        C. Tune Parameters/Hyperparameters of the model/s. [4 Marks]
        

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rfclf2 = RandomForestClassifier(max_depth=200)
rfclf2.fit(x_train_tv, y_train)
y_pred_class = nb.predict(x_test_ct)
print("Number of Features : ", x_train_tv.shape[1])
print("Training Accuracy : ",rfclf2.score(x_train_tv,y_train ))
print("Testing Accuracy : ", rfclf2.score(x_test_tv,y_test))

        D. Clearly print Performance Metrics. [2 Marks]
            Hint: Accuracy, Precision, Recall, ROC-AUC

### 5. Share insights on relative performance comparison [8 Marks]
        A. Which vectorizer performed better? Probable reason?.[2 Marks]


        B. Which model outperformed? Probable reason? [2 Marks]
        

        C. Which parameter/hyperparameter significantly helped to improve performance?Probable reason?. [2 Marks]

        D. According to you, which performance metric should be given most importance, why?. [2 Marks]

> # PART B - 20 Marks

### • DOMAIN: 
    Customer support
### • CONTEXT: 
    Great Learning has a an academic support department which receives numerous support requests every day throughout the year. Teams are spread across geographies and try to provide support round the year. Sometimes there are circumstances where due to heavy workload certain request resolutions are delayed, impacting company’s business. Some of the requests are very generic where a proper resolution procedure delivered to the user can solve the problem. Company is looking forward to design an automation which can interact with the user, understand the problem and display the resolution procedure [ if found as a generic request ] or redirect the request to an actual human support executive if the request is complex or not in it’s database.
### • DATA DESCRIPTION: 
    A sample corpus is attached for your reference. Please enhance/add more data to the corpus using your linguistics skills.
### • PROJECT OBJECTIVE: 
    Design a python based interactive semi - rule based chatbot which can do the following:
        1. Start chat session with greetings and ask what the user is looking for. [5 Marks]
        2. Accept dynamic text based questions from the user. Reply back with relevant answer from the designed corpus. [10 Marks]
        3. End the chat session only if the user requests to end else ask what the user is looking for. Loop continues till the user asks to end it. [5 Marks]
        Hint: There are a lot of techniques using which one can clean and prepare the data which can be used to train a ML/DL classifier. Hence, it might require you to experiment, research, self learn and implement the above classifier. There might be many iterations between hand building the corpus and designing the best fit text classifier. As the quality and quantity of corpus increases the model’s performance i.e. ability to answer
        right questions also increases.
        Reference: https://www.mygreatlearning.com/blog/basics-of-building-an-artificial-intelligence-chatbot/
### • Evaluation: 
    Evaluator will use linguistics to twist and turn sentences to ask questions on the topics described in DATA DESCRIPTION and check if the bot is giving relevant replies

In [20]:
texttospeech

NameError: name 'texttospeech' is not defined

In [ ]:
import pyttsx3
text = pyttsx3.init()
s = 'The code is Compiled'
text.say(s)
text.runAndWait()